### DISCUSSION TITLE
beta.55 Improved Search, Drag, and Happiness 😊

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#405] [Announcements]
## Search Improvements

beta.55 has a vastly improved search:

Search descriptions 🎉

![CleanShot 2021-08-20 at 13 37 44](https://user-images.githubusercontent.com/36073/130285547-24f111d7-a706-4be2-b0d6-b425afbe6683.png)

Search shortcuts

![CleanShot 2021-08-20 at 13 51 49](https://user-images.githubusercontent.com/36073/130286634-4797c029-c2ed-4071-9e1d-285d6bf1a15f.png)

Search by kenv

![CleanShot 2021-08-20 at 13 51 18](https://user-images.githubusercontent.com/36073/130286567-4fd0a155-43a7-4af5-bd3f-c0a9db9ae8dc.png)

Sear by "command-name" (if you can't think of // Menu: name)

![CleanShot 2021-08-20 at 13 54 45](https://user-images.githubusercontent.com/36073/130286878-62b2a139-b3b7-4c34-904f-40d7b03a7e1c.png)

Sorts by "score" (rather than alphabetically)

## Drag

Choices can now take a `drag` property. This will make list items "draggable" and allow you to drag/drop to copy files from your machine (or even from URLs!) into any app. When using remote URLs, their will be a bit of "delay" while the file downloads (depending on the file size) between "drag start" and "drop enabled", so just be aware. I'll add some sort of download progress indicator sometime in the future, just not high priority 😅

```js
// Menu: Drag demo

await arg(
  {
    placeholder: "Drag something from below",
    ignoreBlur: true,
  },
  [
    {
      name: "Heart Eyes (local)",
      drag: "/Users/johnlindquist/Downloads/john-hearts@2x.png",
      img: "/Users/johnlindquist/Downloads/john-hearts@2x.png",
    },
    {
      name: "React logo svg (wikipedia)",
      drag: "https://upload.wikimedia.org/wikipedia/commons/a/a7/React-icon.svg",
      img: "https://upload.wikimedia.org/wikipedia/commons/a/a7/React-icon.svg",
    },
  ]
)
```

![CleanShot 2021-08-20 at 15 26 07](https://user-images.githubusercontent.com/36073/130294979-c45aabe2-6c30-41ad-94b4-64a85a2c34eb.gif)

You can use the `drag` object syntax to define a `format` and `data`

> `text/html`: Renders the HTML payload in contentEditable elements and rich text (WYSIWYG) editors like Google Docs, Microsoft Word, and others.
> `text/plain`: Sets the value of input elements, content of code editors, and the fallback from text/html.
> `text/uri-list`: Navigates to the URL when dropping on the URL bar or browser page. A URL shortcut will be created when dropping on a directory or the desktop.

```js
// Menu: Drag demo

await arg(
  {
    placeholder: "Drag something from below",
    ignoreBlur: true,
  },
  [
    {
      name: "Padding 4",
      drag: {
        format: "text/plain",
        data: `className="p-4"`,
      },
    },
    {
      name: "I love code",
      drag: {
        format: "text/html",
        data: `<span style="background-color:yellow;font-family:Roboto Mono">I ❤️ code</span>`,
      },
    },
  ]
)
```

![CleanShot 2021-08-20 at 15 48 00](https://user-images.githubusercontent.com/36073/130296713-6249d5c2-c01f-42d1-b2c2-ea86d2e4c29b.gif)

## Happiness

I'm _very_ happy with the state of Script Kit. When I started almost a year ago, I had no idea I could push the concept of creating/sharing/managing custom scripts so far. I think it looks great, feels speedy, and is flexible enough to handle so, so many scenarios.

With everything in place, next week I'm starting on creating lessons, demos, and docs. It's time to show you what Script Kit can really do 😉 

P.S. - Thanks for all the beta-testing and feedback. It's been tremendously helpful!

**Comment by hakubo** [comment]
One thing I noticed after the upgrade is it changed how things are sorted! So my memory muscle is bit lost now.
I will probably get used to it but have found one thing that you might look into:
![Screenshot 2021-08-23 at 08 41 11](https://user-images.githubusercontent.com/1018759/130402300-e47b3306-e7e0-466e-977a-60ec51b7667b.png)

Shouldn't Google Chrome be higher as it can match a substring in it?

**Reply by johnlindquist** [reply]
I see what you're saying. I'm using a lib called [quick-score](https://www.npmjs.com/package/quick-score).

I'm sticking with the default "Quicksilver" fuzzy-match algo for now: https://github.com/fwextensions/quick-score/blob/master/src/quick-score.js#L55

The first result scores so high because of how it treats "skipped characters". `Ch` is 0 skipped characters, so that's MAX score, then the "r" is only 1 character away, which is a tiny deduction. `Chr` in `Google Chrome` has to skip 7 characters just to make the first match.

So the concept of "exact substring" doesn't really apply. It's always "one char a time, weighted by distance from last match".

I imagine I could apply `'Google Chrome'.match(/Chr(?<=(^|\w))/i)` after QuickSilver on the sort so that exact matches following start/whitespace/dash get pulled to the top. I'll need to think about it and test it a bit.

**Reply by johnlindquist** [reply]
Ok, I added "lookbehinds" to score super high in the algo:

Ranked like so:
1. "starts with"
2. Preceded by whitespace
3. Preceded by dash
4. Preceded by non-word char
5. Default back to Quicksilver algo

```js
const precede = [`(?<=^)`, `(?<=\\s)`, `(?<=-)`, `(?<=\\W)`];
function kitScorer(string: string, query: string, matches: number[][]) {
  for (let i = 0; i < precede.length; i++) {
    const p = precede[i];
    const r = new RegExp(`${p}${query}`, 'i');
    const match = string.match(r);
    if (match) {
      const start = match?.index || 0;
      const { length } = match[0];
      const ms = [start, start + length];
      matches.push(ms); //where to highlight chars
      return 1 - i / 100;
    }
  }

  return quickScore(string, query, matches);
}
```

The result looks like this:
![CleanShot 2021-08-23 at 13 02 28](https://user-images.githubusercontent.com/36073/130503074-f9e69fe5-2454-4a11-b260-8e3224650c24.png)

Pretty happy with it so far. Will push out a build tonight/tomorrow.
